_posts/2019-07-15-writing-fast-rails-part-3.markdown

```diff
@@ -0,0 +1,58 @@
+---
+layout: post
+title:  "Tips for Writing Fast Rails: Part 3"
+date: 2019-07-15 08:00
+categories: ["rails", "performance"]
+author: "luciano"
+---
+
+Here we continue with the [series](https://fastruby.io/blog/tags/performance) of articles where we talk how minor adjustments in the code can lead to major performance improvements.
+
+In this article we'll focus on the use of [ActiveRecord::Batches#find_each](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each) when it comes to iterations across a large number of records.
+
+<!--more-->
+
+## Prefer `Model.find_each` instead of `Model.all.each`
+
+A very common mistake (often seen in background jobs) is to use [ActiveRecord::Scoping::Named::ClassMethods#all](https://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html#method-i-all) + [ActiveRecord::Result#each](https://api.rubyonrails.org/classes/ActiveRecord/Result.html#method-i-each) (`Model.all.each`) to iterate a table with several thousands of records on it.
+
+```
+Product.all.each do |product|
+  product.update_column(:stock, 50)
+end
+```
+
+The problem with that code is that `Product.all` initialize all the records at once and loads them into memory. This can lead to major memory usage issues in your app.
```

Comment:
```suggestion
The problem with that code is that `Product.all` initializes all the records at once and loads them into memory. This can lead to major memory usage issues in your app.
```